pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -205,6 +206,18 @@ private static void removeDocId(IndexSegment segment, int docId) {
   @Override
   protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
     assert !validDocIds.isEmpty();
+
+    // Skip removing segments that has segment EndTime in the comparison cols earlier than (largestSeenTimestamp - TTL).
```

Comment:
This part should be moved to the `BasePartitionUpsertMetadataManager.removeSegment(ImmutableSegment segment)` before locking the snapshot read lock